### ISSUE TITLE
Rename `concat_str` to `join_horizontal` for consistency

### ISSUE THREAD

**Issue by JulianCologne** [pola-rs/polars#12469]
### Description

`concat_str` should be named `join_horizontal` for consistency.
- many functions that work horizontally in polars have the suffix `_horizontal` (`min_horizontal`, `sum_horizontal`, `all_horizontal`, ...)
- `join` is used in python and rust to flatten data with a separator: `" ".join(["hello", "world"])` (python); `["hello", "world"].join(" ")` (rust)
- polars uses `join` in `Expr.list.join` and `Series.list.join` already so it is inconsisten within polars

**Comment by alexander-beedie** [comment]
Hmm, not sure about this; the name `join` at that level feels like it collides with join _operations_ (which the `list.join` doesn't as it's in a very specific namespace).

Functionally-speaking it is also _not_ equivalent to the other `join` expressions mentioned, which join values in a single Series with a separator - `concat_str` concatenates an arbitrary number of _expressions_, which it can further (optionally) join with a separator.

In other words, converging the names would actually be the inconsistent choice, given that these functions do different things 🤔

**Comment by cmdlineluser** [comment]
It almost feels like `.list.join()` should be called `.list.to_string()` instead.

**Comment by JulianCologne** [comment]
> Hmm, not sure about this; the name `join` at that level feels like it collides with join _operations_ (which `list.join` doesn't as it exists within a specific expression namespace).

I dont think this is a problem since we have a very different context:
- join DataFrames: `df.join(df_other, on="idx")`
- join data into text: `df.with_columns(pl.join_horizontal("a", "b", separator="-"))`




> Functionally-speaking it is also _not_ equivalent to the other `join` expressions mentioned

I disagree ;)
`join` is a function that combines/flattens data into a string with a separator in between.
Functionality-wise they are exactly the same. The polars user wants to join data vertical/horizontal and does not care about some implementation details.
Implementation-wise they are different, yes! But the API is created for users and they might not expect different names because the underlaying implementation works a bit different.




> It almost feels like `.list.join()` should be called `.list.to_string()` instead.

Almost every language uses `join() + separator` to flatten data into text and some kind of `to_string()` without any parameter to convert to text so I would definitely not change that.

**Comment by deanm0000** [comment]
> Almost every language uses join() + separator to flatten data into text and some kind of to_string() without any parameter to convert to text so I would definitely not change that.

yeah but they don't have join to mean anything else. I think the better analog would be how database engines do it. Postgres, duckdb, and sql server (those are the only 3 I looked at) all have concat_ws.

Excel uses concat. R uses `paste0`.

**Comment by cmdlineluser** [comment]
> without any parameter

I did mean for the parameter to still exist e.g. `.list.to_string(separator=", ")`

e.g. similar to `array_to_string(array, separator)` in SQL #10858

**Comment by alexander-beedie** [comment]
I'm going to close this one out as it feels like there isn't really a strong win here, whereas there _is_ potential confusion over a top-level "join" function that definitely doesn't do the same thing as the frame-level "join" functions ;) 
`concat_str` isn't unclear, and there is plenty of precedent for the naming from SQL/etc.
